### PR TITLE
Query settings through REST API

### PR DESCRIPTION
**Pull request by joaquimvila**
* Added new **JSONSettingsRESTAPI** library to facilitate the creation of a REST API for settings handling
* Implemented the following settings **endpoints**:
** Get all available settings of a file
** Query the details for a particular setting (given its identifier)
* Packed new library as a conan package
* Added automated tests for the endpoints integrated on the library
* Added tests for the new conan package
* Updated CI environments (appveyor & travis) to also build/deploy the new library

### CODE REVIEW

File: src/JSONSettingsRESTAPI/Endpoints/SettingsGetEndpoint.cpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+		auto settings = SettingDefinitionMgr::get().getFileSettings(m_settingsFile);
+		for (const auto& settingData : settings)
+		{
+			if (settingData.second.id == settingId)
+			{
+				return settingData.first;
+			}
+		}
+
+		return boost::none;
```
Comment (PacoGutierrez) [inline src/JSONSettingsRESTAPI/Endpoints/SettingsGetEndpoint.cpp:76]: How can it affect that there are no settings?
Is the App protected against this?

File: src/JSONSettingsRESTAPI/Helpers/ReplyBuilderHelper.cpp [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	ReplyBuilderHelper::buildEmpty(systelab::web_server::Reply::StatusType status)
+	{
+		return build(status, "{}");
+	}
+
+	std::unique_ptr<systelab::web_server::Reply>
+	ReplyBuilderHelper::buildMessage(systelab::web_server::Reply::StatusType status, const std::string& message)
+	{
+		std::string cleanMessage = message;
+		std::replace(cleanMessage.begin(), cleanMessage.end(), '\"', '\'');
```
Comment (PacoGutierrez) [inline src/JSONSettingsRESTAPI/Helpers/ReplyBuilderHelper.cpp:19]: Why is this replacement performed?

File: src/JSONSettingsRESTAPI/Helpers/SettingCurrentValueHelper.cpp [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+	SettingValue SettingCurrentValueHelper::query(const ISettingsService& service,
+												  const SettingsFile& file,
+												  const SettingPath& path,
+												  const SettingValueType& valueType)
+	{
+		SettingValue currentValue;
+		if (valueType == SettingValueType::IntValue)
+		{
+			currentValue = SettingValue(service.getSettingInteger(file, path)).value;
```
Comment (PacoGutierrez) [inline src/JSONSettingsRESTAPI/Helpers/SettingCurrentValueHelper.cpp:17]: How these types are managed for a single value within the schema

File: src/JSONSettingsRESTAPI/Helpers/SettingCurrentValueHelper.h [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+namespace systelab { namespace setting {
+	class ISettingsService;
+}}
+
+namespace systelab { namespace setting { namespace rest_api {
+
+	class SettingCurrentValueHelper
+	{
+	public:
+		static SettingValue query(const ISettingsService&, const SettingsFile&, const SettingPath&, const SettingValueType&);
```
Comment (PacoGutierrez) [inline src/JSONSettingsRESTAPI/Helpers/SettingCurrentValueHelper.h:17]: Why is it necessary that this variable to be static? Is it used out of this scope?

File: src/JSONSettingsRESTAPI/test_package/JSONSettingsRESTAPIExample.cpp [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+int main(int argc, char *argv[])
+{
+	systelab::setting::SettingsFile settingsFile = "MySettingsFile.json";
+	auto settingsService = std::make_unique<systelab::setting::SettingsService>();
+	systelab::json::rapidjson::JSONAdapter jsonAdapter;
+	
+	auto endpoint = std::make_unique<systelab::setting::rest_api::SettingsGetAllEndpoint>
+						(settingsFile, std::move(settingsService), jsonAdapter);
+	endpoint->execute(systelab::rest_api_core::EndpointRequestData());
```
Comment (PacoGutierrez) [inline src/JSONSettingsRESTAPI/test_package/JSONSettingsRESTAPIExample.cpp:30]: What is the objective for this example? What should we expect?

File: src/JSONSettingsRESTAPI/Helpers/ReplyBuilderHelper.cpp (hunk 2 above)
Comment (joaquimvila) [reply src/JSONSettingsRESTAPI/Helpers/ReplyBuilderHelper.cpp:19]: It is made to sanitize the possible conflictive chars contained on the string message, such as quotes or carry returns. This way, we guarantee that the reply JSON is well-formed.

File: src/JSONSettingsRESTAPI/Endpoints/SettingsGetEndpoint.cpp (hunk 1 above)
Comment (joaquimvila) [reply src/JSONSettingsRESTAPI/Endpoints/SettingsGetEndpoint.cpp:76]: If there are no settings, the endpoint will reply with a "Not found" reply. 
Moreover, if you check where this function is called (in the execute() method of the endpoint), then you will see that the endpoint is doing this handling when the optional hasn't a value .

File: src/JSONSettingsRESTAPI/Helpers/SettingCurrentValueHelper.cpp (hunk 3 above)
Comment (joaquimvila) [reply src/JSONSettingsRESTAPI/Helpers/SettingCurrentValueHelper.cpp:17]: All types are serialized as a string inside the SettingValue struct. Morevoer, we keep track of the specific type with another type attribute (inside the SettingValue struct).

File: src/JSONSettingsRESTAPI/Helpers/SettingCurrentValueHelper.h (hunk 4 above)
Comment (joaquimvila) [reply src/JSONSettingsRESTAPI/Helpers/SettingCurrentValueHelper.h:17]: It is not the variable but the method who is static. 
It is made this way to facilitate the usage of this helper from the endpoint: using static methods allow not creating an instance, so the code needed to use them is reduced.

File: src/JSONSettingsRESTAPI/test_package/JSONSettingsRESTAPIExample.cpp (hunk 5 above)
Comment (joaquimvila) [reply src/JSONSettingsRESTAPI/test_package/JSONSettingsRESTAPIExample.cpp:30]: This is just a smoke test of the conan package. It is executed when using the "conan test" command.

It allows checking that the conan package for this library has been generated successfully.
With this simple file, it check that includes are on the right folder and that the application can link properly, so all the symbols needed are also on the package.

Bottom line, it is a good practice when working with conan.

**Review by PacoGutierrez — APPROVED** [review]
Approve changes

**Pull request merged**